BusinessLogic/Implementations/RolesAndPermissionsManager.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using BusinessLogic.Interfaces;
+using Models.DataModels;
+using Repository.ExtendedRepositories;
+using Services;
+
+namespace BusinessLogic.Implementations
+{
+    public class RolesAndPermissionsManager : IRolesAndPermissionsManager
+    {
+        private readonly IPermissionsRepository _permissionsRepo;
+        private readonly IRolesRepository _rolesRepo;
+        private readonly IUserRepository _userRepo;
+        public RolesAndPermissionsManager(IPermissionsRepository permissionsRepo, IRolesRepository rolesRepo, IUserRepository userRepo)
+        {
+            _permissionsRepo = permissionsRepo;
+            _rolesRepo = rolesRepo;
+            _userRepo = userRepo;
+        }
+        public IQueryable<Role> GetAllRoles()
+        {
+            return _rolesRepo.GetAll();
+        }
+
+        public void InsertRole(Role newRole)
+        {
+            if (Helpers.HasNullOrEmptyStrings(newRole)) throw new Exception("The New Role To be added Contains null values");
+            _rolesRepo.Insert(newRole).Wait();
+        }
+
+        public void DeleteRole(int id)
+        {
+            if (id <= 0) throw new KeyNotFoundException($"id {id} doesn't exist in the Database");
+            var role = _rolesRepo.Get(id);
+            _rolesRepo.SoftDelete(role);
+        }
+
+        public IQueryable<Permission> GetAllPermissions()
+        {
+            return _permissionsRepo.GetAll();
+        }
+
+        public void InsertPermission(Permission newPermission)
+        {
+            if (Helpers.HasNullOrEmptyStrings(newPermission)) throw new Exception("The New Permission To be added Contains null values");
+            _permissionsRepo.Insert(newPermission).Wait();
+        }
+
+        public void DeletePermission(int id)
+        {
+            if (id <= 0) throw new KeyNotFoundException($"id {id} doesn't exist in the Database");
+            var permission = _permissionsRepo.Get(id);
+            _permissionsRepo.SoftDelete(permission);
+        }
+
+        public IQueryable<Permission> GetPermissionsOfRole(string roleName)
+        {
+            var role = _rolesRepo.GetRole(roleName);
+            return role != null
+                ? _permissionsRepo.GetPermissionsOfRole(role.Id)
```

Comment:
GetPermissionsOfRole has an overload that takes a role string directly, use that for optimization

Reply:
this is done for validation purposes to make sure that the roles exists

Reply:
validate by handling exceptions thrown from the database instead of doubling/tripling the amount of required quiries to finish the process